### DISCUSSION TITLE
How to define an abstract property?

### DISCUSSION THREAD

**Discussion by GregDungca** [pydantic/pydantic#2410] [Question] [answered]
I'd like to define a base class that specifies a set of fields that children of this class must define.

I expected that I'd be able to do this via abstract base classes, but that's not the case as seen below. 

Best workaround I can come up with is to write unit tests that run on CI which fail if the abstract property isn't defined. Not at all ideal. Ideas appreciated!
```py
import abc

from pydantic import BaseModel


class Fruit(BaseModel, abc.ABC):
    @property
    @abc.abstractmethod
    def type(self) -> str:
        """The name of the type of fruit."""


class Apple(Fruit):
    type: str = "apple"
```
```
NameError                                 Traceback (most recent call last)
<ipython-input-1-fc57530c46af> in <module>
     11
     12
---> 13 class Apple(Fruit):
     14     type: str = "apple"

/fn/lib/python3.8/site-packages/pydantic/main.cpython-38-x86_64-linux-gnu.so in pydantic.main.ModelMetaclass.__new__()

/fn/lib/python3.8/site-packages/pydantic/utils.cpython-38-x86_64-linux-gnu.so in pydantic.utils.validate_field_name()

NameError: Field name "type" shadows a BaseModel attribute; use a different field name with "alias='type'".
```

Is this currently possible or would this be a feature request? 

Related issue is here: https://github.com/samuelcolvin/pydantic/issues/2334

**Comment by samuelcolvin** [comment, marked as answer]
```py
import abc
from typing import ClassVar

from pydantic import BaseModel
from devtools import debug

class Fruit(BaseModel, abc.ABC):
    @property
    @abc.abstractmethod
    def type(self) -> str:
        """The name of the type of fruit."""

class Apple(Fruit):
    type: ClassVar[str] = "apple"
    size: int

a = Apple(size=4)
debug(a, a.type)
```

Unless you want to allow `type` to be changed when creating a model instance, in which case just make it a field on `Fruit`.

**Reply by alexvoronov** [reply]
Is there a way to combine it with `Literal` to separate different classes? 

I tried the following modification of the code, but get validation error "Can't instantiate abstract class Apple with abstract method type (type=type_error)":
```python
import abc
from typing import ClassVar, Literal, Union

from pydantic import BaseModel

class Fruit(BaseModel, abc.ABC):
    @property
    @abc.abstractmethod
    def type(self) -> str:
        """The name of the type of fruit."""

class Apple(Fruit):
    type: ClassVar[Literal["apple"]]
    size: int

class Banana(Fruit):
    type: ClassVar[Literal["banana"]]
    size: int

AnyFruit = Union[Apple, Banana]

class FruitBasket(BaseModel):
    fruit: AnyFruit

b = FruitBasket(fruit={"type": "banana", "size": 4})
# pydantic.error_wrappers.ValidationError: 2 validation errors for FruitBasket
# fruit
#   Can't instantiate abstract class Apple with abstract method type (type=type_error)
# fruit
#   Can't instantiate abstract class Banana with abstract method type (type=type_error)

```

I tried another modification below, and I get no validation error, but the `Fruit` is always `Apple`, even if it says `type='banana'`:
```python
import abc
from typing import ClassVar, Literal, Union

from pydantic import BaseModel

class Fruit(BaseModel, abc.ABC):
    type: str

class Apple(Fruit):
    type: ClassVar[Literal["apple"]]
    size: int

class Banana(Fruit):
    type: ClassVar[Literal["banana"]]
    size: int

AnyFruit = Union[Apple, Banana]

class FruitBasket(BaseModel):
    fruit: AnyFruit

b = FruitBasket(fruit={"type": "banana", "size": 4})
# >>> b
# FruitBasket(fruit=Apple(type='banana', size=4))
```

**Reply by ghost** [reply]
Thank you @samuelcolvin , exactly what I was looking for! 

Can you help with one extra step? How can I create a validator for the abstract field in the base class? I tried the obvious thing but it doesn't work. The body of the validator doesn't run, even `assert False` doesn't trigger.

```python
class Fruit(BaseModel, abc.ABC):
    @property
    @abc.abstractmethod
    def type(self) -> str:
        """The name of the type of fruit."""
        
    @validator('type', check_fields=False)  # got an error without this
    def type_must_be_tropical(cls, t):
        assert t in ['pineapple', 'coconut', 'guava']
        return t

class Apple(Fruit):
    type: ClassVar[str] = "apple"
    size: int

a = Apple(size=4)
debug(a, a.type)
```

**Comment by PrettyWood** [comment]
Hi @AlexanderSov @GregDungca 
IMO the easiest solution is to use `__init_subclass__` to check if all fields are well defined
For example
```py
from typing import Literal

from pydantic import BaseModel
from pydantic.typing import is_literal_type, all_literal_values

_all_type_values: set[str] = set()

class Fruit(BaseModel):
    def __init_subclass__(cls, **kwargs):
        try:
            type_ = cls.__fields__['type'].outer_type_
        except KeyError:
            raise TypeError(f"Class {cls.__name__!r} needs field 'type'")

        if not is_literal_type(type_):
            raise TypeError(f"Class {cls.__name__!r} needs the field 'type' to be a `Literal`")

        type_values = set(all_literal_values(type_))

        global  _all_type_values

        if _all_type_values.intersection(type_values):
            raise ValueError(f"Some values of 'type' {type_!r} are already taken")

        _all_type_values |= type_values

class Apple(Fruit):
    type: Literal["banana"] = "apple"
    size: int

class Banana(Fruit):
    type: Literal["banana"] = "banana"
    size: int
```

Hope this helps!

**Reply by PrettyWood** [reply]
Btw https://github.com/samuelcolvin/pydantic/pull/2336 will probably help as this check is more or less [done for you](https://github.com/samuelcolvin/pydantic/pull/2336/files#diff-bfaaac2ae0a71d9686a272982d29d0bed82bfaacf8b5b30e2b88ed2543e74a3bR1094) automatically :)

**Reply by coltonbh** [reply]
@PrettyWood can you point me to a simple way to express the idea above? I'm not seeing in it #2336 clearly... I'm new to discriminators so not totally following the plot. Thanks!

**Reply by frederickch** [reply]
Something like this?

```python
import abc
from typing import Literal, Union

from pydantic import BaseModel, Field

class Fruit(BaseModel, abc.ABC):
    type: str

class Apple(Fruit):
    type: Literal["apple"]
    size: int

class Banana(Fruit):
    type: Literal["banana"]
    size: int

AnyFruit = Union[Apple, Banana]

class FruitBasket(BaseModel):
    fruit: AnyFruit = Field(discriminator="type")

```

**Reply by coltonbh** [reply]
Thanks! Two questions:

1. What's the `type` field really doing anyways? It's it a duplicate piece of information that is basically the same thing as `__class__`. The `class` defines the type. Why put that in string form? It's at `__class__.__name__` if needed. (Clearly there's a reason, so I'm asking for enlightenment)
2. What is the discriminator really discriminating? If `Union[Apple, Banana]` is OK, seems there's nothing left to discriminate since each class defines the `type: Literal` field so each `Banana` will have `.type="banana"`; there's no real validation to be done. Again, I think I may be missing the plot in some way...

Thanks for your help!

**Reply by Drino** [reply]
@coltonbh 

1. It seems to actually store `type` field on any instance. It's kind of strange to have `__class__` field in your JSON Schema (actually, dunder-methods are pretty python-specific), so usually there is some `type` or even `@type` field in JSON object to discriminate between data structures. 
2. Unions in pydantic are [pretty straightforward](https://pydantic-docs.helpmanual.io/usage/types/#unions) - for field with type `Union[str, int]` pydantic even casts number `42` to string `'42'`. So "Discriminated Unions" discriminate between types in union using some hint.

I'm not sure if discriminated union is actually the thing you are looking for, as in [docs](https://pydantic-docs.helpmanual.io/usage/types/#literal-type) there are examples where discrimination between union types happens automatically based on `Literal[]` field value.

**Reply by coltonbh** [reply]
Got it! Makes sense now. The discriminator is making explicit which object in the `Union` the data should be cast as. Thanks for the `Union[str, int]` example--that cleared it up :)

**Comment by coltonbh** [comment]
Is there a best practice way to express the idea of "abstract attribute" in pydantic? @GregDungca did you ever discover the correct way to express this idea? Thanks!

**Reply by GregDungca** [reply]
Ah apologies for not responding to this discussion. I didn't end up using the solutions that samuelcolvin or PrettyWood posted, but they do look like decent workarounds.

I was able to solve my specific issue, which is not a solution to the general problem, but I'll post it below in case it helps you or others:

Basically what I wanted to do was for a class, define a field for which subclasses would be required to specify a custom default value. In my example, `type` was that field. Again this is very specific to my use case, but what I really wanted was the value for `type` to be some identifier for that subclass. I was able to work around this with the following code where I used the class name as a default value.
```python
import abc

from pydantic import BaseModel
from typing import Optional, Any


class Fruit(BaseModel, abc.ABC):

    type: Optional[str]

    def __init__(self, **data: Any) -> None:
        super().__init__(**data)
        self.type = self.__class__.__name__


class Apple(Fruit):
    pass

a = Apple()
print(a.type)

```

**Reply by coltonbh** [reply]
Oh very nice! That's similar to a problem I had recently where I wanted to use the new discriminator interface for `pydantic` but found adding `type` kind of silly because `type` is essentially defined by the class. Maybe making `.type` private can give me this interface but without exposing a `.type` property that is a duplicate of classname.

As I've thought more about "abstract types" maybe the concept doesn't make sense? Just put the type on the parent class and then it obviously will exist on children classes too--they don't need to implement it?? In my case I wanted to define an interface where children must specify more specific types on the "abstract fields" (very similar to what you're doing here) so perhaps there's  a way I can adapt this :)

Thank you!
